tests/components/media_player/test_kef.py

```diff
@@ -0,0 +1,234 @@
+        if not self.online:
+            raise ConnectionRefusedError("Offline")
+        if not self.__muted:
+            self.__volume = self.__volume + (step if step else 0.05)
+
+    # pylint: disable=invalid-name
+    def decreaseVolume(self, step=None):
+        """Mock for decrease volume."""
+        self.increaseVolume(-(step or 0.05))
+
+
+def add_entities_factory(self):
+    """Add devices factory."""
+    def add_entities(devices, update_befor_add=False):
+        """Fake add device."""
+        self.devices += devices
+
+    return add_entities
+
+
+class TestKefMediaPlayer(unittest.TestCase):
+    """Unit tests for kef component."""
+
+    # pylint: disable=invalid-name
+    def setUp(self):
+        """Set up things to be run when tests are started."""
+        self.hass = get_test_home_assistant()
+        self.config = {
+            DOMAIN: {
+                CONF_PLATFORM: 'kef',
+                CONF_HOST: '192.0.1.1',
+            }
+        }
+        self.turn_on_config = {
+            DOMAIN: {
+                CONF_PLATFORM: 'kef',
+                CONF_HOST: '192.0.1.1',
+                CONF_NAME: 'speaker',
+                CONF_TURN_ON_SERVICE: "homeassistant.turn_on",
+                CONF_TURN_ON_DATA: "{}"
+            }
+        }
+        self.devices = list()
+
+    # pylint: disable=invalid-name
+    def tearDown(self):
+        """Stop everything that was started."""
+        self.hass.stop()
+
+    @mock.patch('pykef.KefSpeaker', side_effect=KefSpeakerMock)
+    def test_setup(self, kef_speaker):
+        """Test a single device setup."""
+        setup_component(self.hass, DOMAIN, self.config)
+        devices = list(self.hass.data[DOMAIN].entities)
+        self.assertEqual(len(devices), 1)
+        kef_speaker.assert_called_once_with('192.0.1.1', 50001)
+
+    @mock.patch('pykef.KefSpeaker', side_effect=KefSpeakerMock)
+    def test_setup_with_turn_on_service(self, kef_speaker):
+        """Test a single device with turn_on service in config."""
+        setup_component(self.hass, DOMAIN, self.turn_on_config)
+        devices = list(self.hass.data[DOMAIN].entities)
+        self.assertEqual(len(devices), 1)
+
+    @mock.patch('pykef.KefSpeaker', side_effect=KefSpeakerMock)
+    def test_setup_while_offline(self, kef_speaker):
+        """Test a single device while offline."""
+        kef_speaker.online = mock.Mock(return_value=False)
+        setup_component(self.hass, DOMAIN, self.config)
+        devices = list(self.hass.data[DOMAIN].entities)
+        self.assertEqual(len(devices), 1)
+        kef_speaker.assert_called_once_with('192.0.1.1', 50001)
+
+    @mock.patch('pykef.KefSpeaker', side_effect=KefSpeakerMock)
+    def test_turn_off_service_call(self, kef_speaker):
+        """Test turn off."""
+        hass_mock = mock.Mock()
+        kef.setup_platform(hass_mock, self.config[DOMAIN],
+                           add_entities_factory(self))
+        self.devices[0].update()
```

Comment:
Fire a time changed event while patching time, to have home assistant schedule an update. See the manual alarm control panel tests for example.